From 4ce0017a373afaaa9ef17614d8fa4f6fde261d18 Mon Sep 17 00:00:00 2001
From: Edward Cree <ecree@solarflare.com>
Date: Mon, 2 Jul 2018 16:13:40 +0100
Subject: [PATCH] net: core: another layer of lists, around PF_MEMALLOC skb
 handling
Git-commit: 4ce0017a373afaaa9ef17614d8fa4f6fde261d18
Patch-mainline: v4.19-rc1
References: bsc#1050549
First example of a layer splitting the list (rather than merely taking
 individual packets off it).
Involves new list.h function, list_cut_before(), like list_cut_position()
 but cuts on the other side of the given entry.

[SLE backport: only introduce list_cut_before() api.

Signed-off-by: Edward Cree <ecree@solarflare.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
Signed-off-by: Davidlohr Bueso <dbueso@suse.de>

---
 include/linux/futex.h |  2 +-
 include/linux/list.h  | 30 ++++++++++++++++++++++++++++++
 2 files changed, 31 insertions(+), 1 deletion(-)

diff --git a/include/linux/futex.h b/include/linux/futex.h
index 7c8c931fad99..b6db0aef5747 100644
--- a/include/linux/futex.h
+++ b/include/linux/futex.h
@@ -4,7 +4,7 @@
 #include <linux/ktime.h>
 #include <uapi/linux/futex.h>
 
-struct inode;
+ struct inode;
 struct mm_struct;
 struct task_struct;
 
diff --git a/include/linux/list.h b/include/linux/list.h
index ae537fa46216..85baa403499d 100644
--- a/include/linux/list.h
+++ b/include/linux/list.h
@@ -284,6 +284,36 @@ static inline void list_cut_position(struct list_head *list,
 		__list_cut_position(list, head, entry);
 }
 
+/**
+ * list_cut_before - cut a list into two, before given entry
+ * @list: a new list to add all removed entries
+ * @head: a list with entries
+ * @entry: an entry within head, could be the head itself
+ *
+ * This helper moves the initial part of @head, up to but
+ * excluding @entry, from @head to @list.  You should pass
+ * in @entry an element you know is on @head.  @list should
+ * be an empty list or a list you do not care about losing
+ * its data.
+ * If @entry == @head, all entries on @head are moved to
+ * @list.
+ */
+static inline void list_cut_before(struct list_head *list,
+				   struct list_head *head,
+				   struct list_head *entry)
+{
+	if (head->next == entry) {
+		INIT_LIST_HEAD(list);
+		return;
+	}
+	list->next = head->next;
+	list->next->prev = list;
+	list->prev = entry->prev;
+	list->prev->next = list;
+	head->next = entry;
+	entry->prev = head;
+}
+
 static inline void __list_splice(const struct list_head *list,
 				 struct list_head *prev,
 				 struct list_head *next)
--
2.16.4

